File: src/tokenizer.cpp
#include <tokenizer.h>

#define MIN(a, b) \
    a < b ? a : b

Tokenizer::Tokenizer(std::string source)
{
    this->source = source;
    this->index = 0;
    this->source_size = source.length();
    this->current = source[this->index];
}

void Tokenizer::next()
{
    if (current != '\0' && index < source_size)
    {
        index += 1;
        current = source[index];
    }
}

void Tokenizer::skip_whitespace()
{
    while(current == 13 || current == 10 || current == ' ' || current == '\t')
        next();
}

void Tokenizer::skip_comment()
{

}

char Tokenizer::peek(int buffer)
{
    return source[MIN(index + buffer, source_size)];
}

Token *Tokenizer::move_with_token(Token *token)
{
    next();
    return token;
}

Token *Tokenizer::move_current(Type type)
{
    std::string value = "";
    value += current;

    Token *token = new Token(value, type);
    next();
    return token;
}

Token *Tokenizer::next_token()
{
    while(current != '\0')
    {
        skip_whitespace();

        if(isalpha(current))
            return parse_id();

        if(isdigit(current))
            return parse_number();

        switch(current)
        {
            case '(': return move_current(TOKEN_LEFT_PAREN);
            case ')': return move_current(TOKEN_RIGHT_PAREN);
            case '[': return move_current(TOKEN_LEFT_BRACKET);
            case ']': return move_current(TOKEN_RIGHT_BRACKET);
            case '{': return move_current(TOKEN_LEFT_BRACE);
            case '}': return move_current(TOKEN_RIGHT_BRACE);
            case ',': return move_current(TOKEN_COMMA);
            case ';': return move_current(TOKEN_SEMICOLON);
            case ':': return move_current(TOKEN_COLON);
            case '+': return move_current(TOKEN_PLUS);
            case '-': return move_current(TOKEN_MINUS);
            case '*': return move_current(TOKEN_ASTERISK);
            case '=': return move_with_token(new Token("=", TOKEN_EQUALS));
            case '"': return parse_id();
            case '\0': break;
            default: std::cout << "CURRENT: [" << current << "]" << std::endl; exit(1); break;
        }
    }

    return new Token("0", TOKEN_END_OF_FILE);
}

Token * Tokenizer::parse_id()
{
    std::string value = "";

    if(current == '"')
    {
        next();

        while(current != '"')
        {
            value += current;
            next();
        }

        next();
        return new Token(value, TOKEN_STRING);
    }

    while(isalpha(current))
    {
        value += current;
        next();
    }

    Token *token = nullptr;

    if (value == "if")
        token = new Token(value, TOKEN_IF);
    else if (value == "while")
        token = new Token(value, TOKEN_WHILE);
    else if (value == "def")
        token = new Token(value, TOKEN_LET);
    else if (value == "ret")
        token = new Token(value, TOKEN_RETURN);
    else if (value == "fun")
        token = new Token(value, TOKEN_FUN);
    else
        token = new Token(value, TOKEN_ID);

    return token;
}

Token * Tokenizer::parse_number()
{
    std::string value = "";

    while(isdigit(current))
    {
        value += current;
        next();
    }

    return new Token(value, TOKEN_INT);
}

File: wasm/Tokenizer.cs
using Compiler.Exceptions;

namespace Compiler;

public class Tokenizer
{
    private readonly string _source;
    private readonly int _sourceSize;
    private char _current;
    private int _index;

    public Tokenizer(string source)
    {
        _source = source;
        _index = 0;
        _sourceSize = source.Length;
        _current = source[_index];
    }

    private void Next()
    {
        if (_current != '\0' && _index < _sourceSize)
        {
            _index++;
            _current = _source[_index];
        }
    }

    private void SkipWhitespace()
    {
        if(char.IsWhiteSpace(_current))
            Next();
    }

    private Token? MoveWithToken(Token token)
    {
        if (token == null)
            return null;

        Next();
        return token;
    }

    private Token MoveCurrentType(TokenType type)
    {
        var token = new Token(_current.ToString(), type);
        Next();
        return token;
    }

    public Token NextToken()
    {
        while (_current != '\0')
        {
            SkipWhitespace();

            if (char.IsLetter(_current))
                return ParseId();

            if (char.IsDigit(_current))
                return ParseNumber();

            switch(_current)
            {
                case '(': return MoveCurrentType(TokenType.TOKEN_LPAREN);
                case ')': return MoveCurrentType(TokenType.TOKEN_RPAREN);
                case '{': return MoveCurrentType(TokenType.TOKEN_LBRACE);
                case '}': return MoveCurrentType(TokenType.TOKEN_RBRACE);
                case '+': return MoveCurrentType(TokenType.TOKEN_PLUS);
                case '-': return MoveCurrentType(TokenType.TOKEN_MINUS);
                case '*': return MoveCurrentType(TokenType.TOKEN_STAR);
                case '/': return MoveCurrentType(TokenType.TOKEN_SLASH);
                case '=': return MoveCurrentType(TokenType.TOKEN_EQUALS);
                case ';': return MoveCurrentType(TokenType.TOKEN_SEMICOLON);
                case ':': return MoveCurrentType(TokenType.TOKEN_COLON);
                case ',': return MoveCurrentType(TokenType.TOKEN_COMMA);
                case '"': return ParseId();
                case '\0': break;
                default: throw new UnexpectedCharacterException(); break;
            }
        }

        return new Token("0", TokenType.TOKEN_EOF);
    }

    private Token ParseId()
    {
        var value = string.Empty;

        if (_current == '"')
        {
            Next();

            while (_current != '"')
            {
                value += _current;
                Next();
            }

            Next();
            return new Token(value, TokenType.TOKEN_STRING);
        }

        while (char.IsLetter(_current))
        {
            value += _current;
            Next();
        }

        return value switch
        {
            "if" => new Token(value, TokenType.TOKEN_IF),
            "while" => new Token(value, TokenType.TOKEN_WHILE),
            "var" => new Token(value, TokenType.TOKEN_VAR),
            "fun" => new Token(value, TokenType.TOKEN_FUN),
            "for" => new Token(value, TokenType.TOKEN_FOR),
            "ret" => new Token(value, TokenType.TOKEN_RETURN),
            _ => new Token(value, TokenType.TOKEN_ID),
        };
    }

    private Token ParseNumber()
    {
        var value = string.Empty;

        while(char.IsNumber(_current))
        {
            value += _current;
            Next();
        }

        return new Token(value, TokenType.TOKEN_NUMBER);
    }
}
